# Data Cleaning with Pandas

In [398]:
import pandas as pd

## Scenario

As data scientists, we want to build a model to predict the sale price of a house in Seattle in 2019, based on its square footage. We know that the King County Department of Assessments has comprehensive data available on real property sales in the Seattle area. We need to prepare the data.

### First, get the data!

When working on a project involving data that can fit on our computer, we store it in a `data` directory.

```bash
cd <project_directory>  # example: cd ~/flatiron_ds/pandas-3
mkdir data
cd data
```

Note that `<project_directory>` in angle brackets is a _placeholder_. You should type the path to the actual location on your computer where you're working on this project. Do not literally type `<project_directory>` and _do not type the angle brackets_. You can see an example in the _comment_ to the right of the command above.

Now, we'll need to download the two data files that we need. We can do this at the command line:

```bash
wget https://aqua.kingcounty.gov/extranet/assessor/Real%20Property%20Sales.zip
wget https://aqua.kingcounty.gov/extranet/assessor/Residential%20Building.zip
```

*Note:* If you do not have the `wget` command yet, you can install it: `brew install wget`.

Note that `%20` in a URL translates into a space. Even though you will *never put spaces in filenames*, you may need to deal with spaces that _other_ people have used in filenames.

There are two ways to handle the spaces in these filenames when referencing them at the command line.

#### 1. You can _escape_ the spaces by putting a backslash (`\`, remember _backslash is next to backspace_) before each one:

`unzip Real\ Property\ Sales.zip`

This is what happens if you tab-complete the filename in the terminal. Tab completion is your friend!

#### 2. You can put the entire filename in quotes:

`unzip "Real Property Sales.zip"`

Try unzipping these files with the `unzip` command. The `unzip` command takes one argument, the name of the tile that you want to unzip.


You can use tab completion (press the `tab` key after the first three letters) to fill in the names, including spaces. This will 

In [399]:
sales_df = pd.read_csv('data/EXTR_RPSale.csv')

/Applications/anaconda3/envs/learn-env/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (1,2) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Seeing pink? Warnings are useful!

Note the warning above: `DtypeWarning: Columns (1, 2) have mixed types.` Because we start with an index of zero, the columns that we're being warned about are actually the _second_ and _third_ columns, `sales_df['Major']` and `sales_df['Minor']`.

In [400]:
!pwd

/Users/patrickfuller/flatiron/section01/pandas-3-data-cleaning-seattle-ds-career-040119


In [401]:
!ls

README.md                    pandas-3-data-cleaning.ipynb
data


In [402]:
!cd data/ && pwd
!pwd

/Users/patrickfuller/flatiron/section01/pandas-3-data-cleaning-seattle-ds-career-040119/data
/Users/patrickfuller/flatiron/section01/pandas-3-data-cleaning-seattle-ds-career-040119


In [403]:
sales_df.head()

ExciseTaxNbr   Major Minor DocumentDate  SalePrice    RecordingNbr Volume  \
0       2687551  138860   110   08/21/2014     245000  20140828001436          
1       1235111  664885    40   07/09/1991          0  199203161090      071   
2       2704079  423943    50   10/11/2014          0  20141205000558          
3       2584094  403700   715   01/04/2013          0  20130110000910          
4       1056831  951120   900   04/20/1989      85000  198904260448      117   

  Page PlatNbr PlatType     ...     PropertyType PrincipalUse SaleInstrument  \
0                           ...                3            6              3   
1  001  664885        C     ...                3            0             26   
2                           ...                3            6             15   
3                           ...                3            6             15   
4  053  951120        P     ...                3            0              2   

  AFForestLand  AFCurrentUseLand  AFNonProfitUse  AFHistoricProperty  \
0            N                 N               N                   N   
1            N                 N               N                   N   
2            N                 N               N                   N   
3            N                 N               N                   N   
4            N                 N               N                   N   

  SaleReason PropertyClass SaleWarning  
0          1             8              
1         18             3          11  
2         18             8    18 31 51  
3         11             8    18 31 38  
4          1             9          49  

[5 rows x 24 columns]

In [404]:
sales_df.head().T

0  \
ExciseTaxNbr                                                  2687551   
Major                                                          138860   
Minor                                                             110   
DocumentDate                                               08/21/2014   
SalePrice                                                      245000   
RecordingNbr                                           20140828001436   
Volume                                                                  
Page                                                                    
PlatNbr                                                                 
PlatType                                                                
PlatLot                                                                 
PlatBlock                                                               
SellerName          WENKLE NOEL SMITH -TRUSTEE                    ...   
BuyerName           ALEXANDER APRIL                               ...   
PropertyType                                                        3   
PrincipalUse                                                        6   
SaleInstrument                                                      3   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                          1   
PropertyClass                                                       8   
SaleWarning                                                             

                                                                    1  \
ExciseTaxNbr                                                  1235111   
Major                                                          664885   
Minor                                                              40   
DocumentDate                                               07/09/1991   
SalePrice                                                           0   
RecordingNbr                                           199203161090     
Volume                                                            071   
Page                                                              001   
PlatNbr                                                        664885   
PlatType                                                            C   
PlatLot                                                B102             
PlatBlock                                                               
SellerName          HERRMAN JEFFREY A+PAULINE A                   ...   
BuyerName           HERRMAN PAULINE A TTEE                        ...   
PropertyType                                                        3   
PrincipalUse                                                        0   
SaleInstrument                                                     26   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                         18   
PropertyClass                                                       3   
SaleWarning                                                        11   

                                                                    2  \
ExciseTaxNbr                                                  2704079   
Major                                                          423943   
Minor                                                              50   
DocumentDate                                               10/11/2014   
SalePrice                                          

### Data overload?

That's a lot of columns. We're only interested in identifying the date, sale price, and square footage of each specific property. What can we do?

In [405]:
sales_df = sales_df[['Major', 'Minor', 'DocumentDate', 'SalePrice']]

In [406]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2014336 entries, 0 to 2014335
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.5+ MB


In [407]:
bldg_df = pd.read_csv('data/Residential Building.zip')

/Applications/anaconda3/envs/learn-env/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Another warning! Which column has index 11?

In [408]:
bldg_df.columns[11]

'ZipCode'

`ZipCode` seems like a potentially useful column. We'll need it to determine which house sales took place in Seattle.

In [409]:
bldg_df.head().T

0                            1  \
Major                                    4300                         4610   
Minor                                     167                          399   
BldgNbr                                     1                            1   
NbrLivingUnits                              1                            1   
Address             15223   40TH AVE S  98188  4431   FERNCROFT RD   98040   
BuildingNumber                          15223                        4431    
Fraction                                                                     
DirectionPrefix                                                              
StreetName          40TH                         FERNCROFT                   
StreetType                             AVE                          RD       
DirectionSuffix                            S                                 
ZipCode                                 98188                        98040   
Stories                                     1                            2   
BldgGrade                                   6                            9   
BldgGradeVar                                0                            0   
SqFt1stFloor                              990                          980   
SqFtHalfFloor                               0                            0   
SqFt2ndFloor                                0                         1140   
SqFtUpperFloor                              0                            0   
SqFtUnfinFull                               0                            0   
SqFtUnfinHalf                               0                            0   
SqFtTotLiving                             990                         2950   
SqFtTotBasement                           990                          940   
SqFtFinBasement                             0                          830   
FinBasementGrade                            0                            9   
SqFtGarageBasement                          0                            0   
SqFtGarageAttached                          0                          480   
DaylightBasement                                                         Y   
SqFtOpenPorch                              80                            0   
SqFtEnclosedPorch                           0                            0   
SqFtDeck                                    0                          600   
HeatSystem                                  6                            5   
HeatSource                                  1                            2   
BrickStone                                  0                            0   
ViewUtilization                                                          N   
Bedrooms                                    2                            3   
BathHalfCount                               0                            1   
Bath3qtrCount                               0                            0   
BathFullCount                               1                            2   
FpSingleStory                               0                            0   
FpMultiStory                                0                            0   
FpFreestanding                              0                            2   
FpAdditional                                0                            0   
YrBuilt                                  1922                         1978   
YrRenovated                                 0                            0   
PcntComplete                                0                            0   
Obsolescence                                0                            0   
PcntNetCondition                            0                            0   
Condition                                   3                            4   
AddnlCost                                   0                            0   

                                              2                  

### So many features!

As data scientists, we should be _very_ cautious about discarding potentially useful data. But, today, we're interested in _only_ the total square footage of each property. What can we do?


In [410]:
bldg_df = bldg_df[['Major', 'Minor', 'SqFtTotLiving', 'ZipCode']]

In [411]:
bldg_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511412 entries, 0 to 511411
Data columns (total 4 columns):
Major            511412 non-null int64
Minor            511412 non-null int64
SqFtTotLiving    511412 non-null int64
ZipCode          468372 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB


In [412]:
sales_data = pd.merge(sales_df, bldg_df, on=['Major', 'Minor'])

ValueError: You are trying to merge on object and int64 columns. If you wish to proceed you should use pd.concat

### Error!

Why are we seeing an error when we try to join the dataframes?

<table>
    <tr>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2013160 entries, 0 to 2013159
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.4+ MB</pre></td>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511359 entries, 0 to 511358
Data columns (total 4 columns):
Major            511359 non-null int64
Minor            511359 non-null int64
SqFtTotLiving    511359 non-null int64
ZipCode          468345 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB
</pre></td>
    </tr>
</table>

Review the error message in light of the above:

* `ValueError: You are trying to merge on object and int64 columns.`

In [413]:
sales_df['Major'] = pd.to_numeric(sales_df['Major'])

ValueError: Unable to parse string "      " at position 935519

### Error!

Note the useful error message above:

`ValueError: Unable to parse string "      " at position 936643`

In this case, we want to treat non-numeric values as missing values. Let's see if there's a way to change how the `pd.to_numeric` function handles errors.

In [414]:
# The single question mark means "show me the docstring"
pd.to_numeric?

Here's the part that we're looking for:
```
errors : {'ignore', 'raise', 'coerce'}, default 'raise'
    - If 'raise', then invalid parsing will raise an exception
    - If 'coerce', then invalid parsing will be set as NaN
    - If 'ignore', then invalid parsing will return the input
```

Let's try setting the `errors` parameter to `'coerce'`.

In [415]:
sales_df['Major'] = pd.to_numeric(sales_df['Major'], errors='coerce')

Did it work?

In [416]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2014336 entries, 0 to 2014335
Data columns (total 4 columns):
Major           float64
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: float64(1), int64(1), object(2)
memory usage: 61.5+ MB


It worked! Let's do the same thing with the `Minor` parcel number.

In [417]:
sales_df['Minor'] = pd.to_numeric(sales_df['Minor'], errors='coerce')

In [418]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2014336 entries, 0 to 2014335
Data columns (total 4 columns):
Major           float64
Minor           float64
DocumentDate    object
SalePrice       int64
dtypes: float64(2), int64(1), object(1)
memory usage: 61.5+ MB


Now, let's try our join again.

In [419]:
sales_data = pd.merge(sales_df, bldg_df, on=['Major', 'Minor'])

In [420]:
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
3  138860.0  110.0   06/08/2005          0           1490   98002
4  423943.0   50.0   10/11/2014          0            960   98092

In [421]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1437522 entries, 0 to 1437521
Data columns (total 6 columns):
Major            1437522 non-null float64
Minor            1437522 non-null float64
DocumentDate     1437522 non-null object
SalePrice        1437522 non-null int64
SqFtTotLiving    1437522 non-null int64
ZipCode          1322148 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 76.8+ MB


We can see right away that we're missing zip codes for many of the sales transactions. (1321536 non-null entries for ZipCode is fewer than the 1436772 entries in the dataframe.) 

In [422]:
sales_data.loc[sales_data['ZipCode'].isna()].head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
99   858140.0  376.0   05/22/2012          0            900     NaN
100  858140.0  376.0   11/28/2017          0            900     NaN
152  720319.0  520.0   11/20/2013     699950           2840     NaN
153  720319.0  520.0   09/16/2013          0           2840     NaN
163  894677.0  240.0   12/21/2016     818161           2450     NaN

Because we are interested in finding houses in Seattle zip codes, we will need to drop the rows with missing zip codes.

In [423]:
sales_data = sales_data.loc[~sales_data['ZipCode'].isna(), :]
                       #.   ~ tilda reverses the bool for zips being 'notanumber'
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
3  138860.0  110.0   06/08/2005          0           1490   98002
4  423943.0   50.0   10/11/2014          0            960   98092

# Your turn: Data Cleaning with Pandas

### 1. Investigate and drop rows with invalid values in the SalePrice and SqFtTotLiving columns.

Use multiple notebook cells to accomplish this! Press `[esc]` then `B` to create a new cell below the current cell. Press `[return]` to start typing in the new cell.

In [424]:
sales_data.loc[:, ['SalePrice']].describe()

SalePrice
count  1.322148e+06
mean   2.826850e+05
std    7.451584e+05
min   -6.000000e+02
25%    0.000000e+00
50%    1.600000e+05
75%    3.450000e+05
max    3.000000e+07

In [425]:
sales_data.loc[:, ['SalePrice']].quantile(.35)

SalePrice    22500.0
Name: 0.35, dtype: float64

In [426]:
sales_data = sales_data.loc[sales_data['SalePrice'] > 35000, : ]

In [427]:
sales_data.describe()

Major          Minor     SalePrice  SqFtTotLiving
count  854050.000000  854050.000000  8.540500e+05  854050.000000
mean   453174.026072    1471.146278  4.373544e+05    2105.805352
std    286371.174654    2747.415008  8.899545e+05     948.481727
min        40.000000       1.000000  3.500300e+04       0.000000
25%    208520.000000     115.000000  1.695000e+05    1450.000000
50%    388810.000000     320.000000  2.800000e+05    1950.000000
75%    725400.000000     950.000000  4.550000e+05    2570.000000
max    990600.000000    9689.000000  3.000000e+07   48160.000000

In [428]:
sales_data.loc[sales_data['SqFtTotLiving'] >10, ['SqFtTotLiving']].describe()

SqFtTotLiving
count  854035.000000
mean     2105.842324
std       948.449029
min        70.000000
25%      1450.000000
50%      1950.000000
75%      2570.000000
max     48160.000000

In [429]:
sales_data = sales_data.loc[sales_data['SqFtTotLiving'] > 20, : ]

In [430]:
sales_data.describe()

Major          Minor     SalePrice  SqFtTotLiving
count  854035.000000  854035.000000  8.540350e+05  854035.000000
mean   453172.179263    1471.166230  4.373539e+05    2105.842324
std    286369.713689    2747.434878  8.899600e+05     948.449029
min        40.000000       1.000000  3.500300e+04      70.000000
25%    208520.000000     115.000000  1.695000e+05    1450.000000
50%    388810.000000     320.000000  2.800000e+05    1950.000000
75%    725400.000000     950.000000  4.550000e+05    2570.000000
max    990600.000000    9689.000000  3.000000e+07   48160.000000

In [431]:
sales_data.loc[sales_data['SalePrice'].isna(), : ]

Empty DataFrame
Columns: [Major, Minor, DocumentDate, SalePrice, SqFtTotLiving, ZipCode]
Index: []

In [432]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 854035 entries, 0 to 1437520
Data columns (total 6 columns):
Major            854035 non-null float64
Minor            854035 non-null float64
DocumentDate     854035 non-null object
SalePrice        854035 non-null int64
SqFtTotLiving    854035 non-null int64
ZipCode          854035 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 45.6+ MB


### 2. Investigate and handle non-numeric ZipCode values

Can you find a way to shorten ZIP+4 codes to the first five digits?

What's the right thing to do with missing values?

In [433]:
sales_data['ZipCode'].apply(type).unique()

array([<class 'str'>, <class 'float'>], dtype=object)

In [434]:
sales_data.loc[sales_data['ZipCode'].apply(type) == float, ['ZipCode']].head()


ZipCode
164   98032
386   98042
388   98042
390   98042
391   98042

In [435]:
# Read the error message and decide how to fix it.
# Note: using errors='coerce' is the *wrong* choice in this case.
def is_integer(x):
    try:
        _ = int(x)
    except ValueError:
        return False
    return True

# sales_data = 
sales_data.loc[sales_data['ZipCode'].apply(is_integer) == False, : ]
sales_data.loc[sales_data['ZipCode'].apply(is_integer) == False, : ].head(20)
# sales_data.head()

Major   Minor DocumentDate  SalePrice  SqFtTotLiving     ZipCode
10168  370500.0    70.0   11/02/2004     240000           2080  98042-3001
10169  370500.0    70.0   12/11/2017     384950           2080  98042-3001
10170  370500.0    70.0   10/21/2014     267500           2080  98042-3001
10171  370500.0    70.0   07/30/2013     245000           2080  98042-3001
16528  137380.0   330.0   03/19/2002     850000           3390  98199-3014
16529  137380.0   330.0   04/20/2015    1115000           3390  98199-3014
16530  137380.0   330.0   05/31/1985     286900           3390  98199-3014
40183  182205.0  9096.0   08/28/1985      62950           1020  98031-3173
43081  370500.0    90.0   12/08/2004     239500           2080  98042-3001
43082  370500.0    90.0   09/30/2015     290200           2080  98042-3001
53691  124500.0   777.0   09/16/1999     330000           2390  98033-4917
53692  124500.0   777.0   03/12/1997     289000           2390  98033-4917
53693  124500.0   777.0   12/12/1995     255000           2390  98033-4917
53694  124500.0   777.0   11/30/1994      65259           2390  98033-4917
53695  124500.0   777.0   12/15/1994     164000           2390  98033-4917
87152  132831.0    90.0   07/06/1999     228000           2510  98058-9018
87153  132831.0    90.0   10/07/1998     219950           2510  98058-9018
93627  352290.0    15.0   12/22/2004     411000           1510  98136-1728
93628  352290.0    15.0   08/03/2015     585000           1510  98136-1728
93629  352290.0    15.0   05/08/2000     264600           1510  98136-1728

In [436]:
sales_data['ZipCode'].apply(type).unique()

array([<class 'str'>, <class 'float'>], dtype=object)

In [437]:
sales_data['ZipCode'] = sales_data['ZipCode'].apply(str)

In [438]:
sales_data['ZipCode'].apply(type).unique()

array([<class 'str'>], dtype=object)

In [439]:
def long_to_five(string):
    return string[:5]

In [440]:
sales_data['ZipCode'] = sales_data['ZipCode'].apply(long_to_five)

In [ ]:
def is_five_long(string):
    return len(string) == 5

In [447]:
sales_data = sales_data.loc[sales_data['ZipCode'].apply(is_five_long) == True, : ]

In [441]:
# sales_data['ZipCode'] = sales_data['ZipCode'].apply(int) #will throw an error for containing a dot

In [448]:
def doesnt_contains_dot(text):
    if '.' in text:
        return False
    else:
        return True

In [449]:
sales_data = sales_data.loc[sales_data['ZipCode'].apply(doesnt_contains_dot), : ]

In [450]:
sales_data['ZipCode'] = sales_data['ZipCode'].apply(int)

In [451]:
sales_data['ZipCode'] = sales_data['ZipCode'].apply(str)

In [452]:
sales_data['ZipCode'].apply(len).head()

0     5
1     5
6     5
7     5
11    5
Name: ZipCode, dtype: int64

In [453]:
sales_data['ZipCode'].head()

0     98002
1     98002
6     98092
7     98092
11    98008
Name: ZipCode, dtype: object

### 3. Add a column for PricePerSqFt



In [454]:
sales_data['pricepersqft'] = sales_data['SalePrice'] / sales_data['SqFtTotLiving']

In [455]:
sales_data['pricepersqft'].head()

0     164.429530
1      73.355705
6     100.000000
7     132.812500
11    260.955056
Name: pricepersqft, dtype: float64

### 4. Subset the data to 2019 sales only.

We can assume that the DocumentDate is approximately the sale date.

In [456]:
sales_data['DocumentDate'].head()
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
0   138860.0  110.0   08/21/2014     245000           1490   98002   
1   138860.0  110.0   06/12/1989     109300           1490   98002   
6   423943.0   50.0   07/15/1999      96000            960   98092   
7   423943.0   50.0   01/08/2001     127500            960   98092   
11  403700.0  715.0   07/03/2013     464500           1780   98008   

    pricepersqft  
0     164.429530  
1      73.355705  
6     100.000000  
7     132.812500  
11    260.955056

In [457]:
def ends_with_2019(text):
    return text.endswith('2019')

In [458]:
sales_data = sales_data.loc[sales_data['DocumentDate'].apply(ends_with_2019) == True, : ]
sales_data['DocumentDate'].head()

605     03/06/2019
651     03/13/2019
868     02/06/2019
1063    02/15/2019
1341    03/04/2019
Name: DocumentDate, dtype: object

### 5. Subset the data to zip codes within the City of Seattle.

You'll need to find a list of Seattle zip codes!

In [459]:
seattle_zips = [98101, 98102, 98103, 98104, 98105, 98106, 98107, 98108, 98109, 98112, 98115, 98116, 98117, 98118, 98119, 98121, 98122, 98125, 98126, 98133, 98134, 98136, 98144, 98146, 98154, 98164, 98174, 98177, 98178, 98195, 98199]

In [461]:
def in_seattle(text):
    return int(text) in seattle_zips

In [462]:
in_seattle('98102')

True

In [466]:
seattle_sales_data = sales_data.loc[ sales_data['ZipCode'].apply(in_seattle) , : ]
seattle_sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
651   367890.0   30.0   03/13/2019     923000           2550   98144   
868   350160.0  125.0   02/06/2019     935000           2460   98117   
1063  882790.0  870.0   02/15/2019     890000           2380   98105   
2821  945920.0  125.0   01/29/2019    1750000           1400   98118   
5359  186240.0  105.0   03/19/2019     380000            800   98117   

      pricepersqft  
651     361.960784  
868     380.081301  
1063    373.949580  
2821   1250.000000  
5359    475.000000

### 6. What is the mean price per square foot for a house sold in Seattle in 2019?

Don't just type the answer. Type code that generates the answer as output!

In [467]:
seattle_sales_data['pricepersqft'].mean()

507.8655050563845

In [468]:
seattle_sales_data.describe()

Major        Minor     SalePrice  SqFtTotLiving  pricepersqft
count    1216.000000  1216.000000  1.216000e+03    1216.000000   1216.000000
mean   466683.050987  1268.789474  8.230999e+05    1769.769737    507.865505
std    293914.913616  2278.584711  5.229804e+05     861.410572    398.718468
min       360.000000     5.000000  4.000000e+04     320.000000      9.756098
25%    220750.000000   122.000000  5.465000e+05    1197.500000    363.080862
50%    405910.000000   380.000000  6.990000e+05    1600.000000    459.337935
75%    738822.500000  1181.250000  9.406250e+05    2120.000000    569.953787
max    984230.000000  9571.000000  6.000000e+06    8625.000000   9558.823529